### ISSUE TITLE
ImportError: None of these hashing backends are installed: ['pycryptodome', 'pysha3'].

### ISSUE THREAD

**Issue by HSC-GBT-DEV2** [pyinstaller/pyinstaller#3681]
I'm facing the below error:

OS: Ubuntu
python version installed:3.5.2
pyinstaller: 3.4.dev0+7817d3fac

**Command line error**
root@hscuser-OptiPlex-3050:/home/hscuser/package# dist/wallet/wallet
[INFO ] [Logger ] Record log in /root/.kivy/logs/kivy_18-08-09_23.txt
[INFO ] [Kivy ] v1.10.1
[INFO ] [Python ] v3.5.2 (default, Nov 23 2017, 16:37:01)
[GCC 5.4.0 20160609]
[INFO ] [Factory ] 194 symbols loaded
[INFO ] [Image ] Providers: img_tex, img_dds, img_sdl2, img_pil, img_gif (img_ffpyplayer ignored)
[INFO ] [Text ] Provider: sdl2
[INFO ] [Window ] Provider: sdl2(['window_egl_rpi'] ignored)
[INFO ] [GL ] Using the "OpenGL" graphics system
[INFO ] [GL ] Backend used
[INFO ] [GL ] OpenGL version <b'3.0 Mesa 18.0.5'>
[INFO ] [GL ] OpenGL vendor <b'Intel Open Source Technology Center'>
[INFO ] [GL ] OpenGL renderer <b'Mesa DRI Intel(R) HD Graphics 630 (Kaby Lake GT2) '>
[INFO ] [GL ] OpenGL parsed version: 3, 0
[INFO ] [GL ] Shading version <b'1.30'>
[INFO ] [GL ] Texture max size <16384>
[INFO ] [GL ] Texture max units <32>
[INFO ] [Window ] auto add sdl2 input provider
[INFO ] [Window ] virtual keyboard not allowed, single mode, not docked
[INFO ] [Base ] Start application main loop
LaserWallet Starting
Checking to see if there is an already running glaze...
Glaze starting /usr/bin/glaze --datadir /root/.laser/testnet --testnet --nousb
Connecting to IPC socket /root/.laser/testnet/glaze.ipc
INFO [08-09|11:27:27.932] Maximum peer count PTH=25 LES=0 total=25
INFO [08-09|11:27:27.932] Starting peer-to-peer node instance=glaze/v1.0.0-stable/linux-amd64/go1.10
INFO [08-09|11:27:27.932] Allocated cache and file handles database=/root/.laser/testnet/glaze/chaindata cache=768 handles=512
INFO [08-09|11:27:28.083] Persisted trie from memory database nodes=0 size=0.00B time=3.656µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [08-09|11:27:28.084] Initialised chain configuration config="{ChainID: 12 Homestead: 0 DAO: DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: Blaze: 0 Engine: ethash}"
INFO [08-09|11:27:28.084] Disk storage enabled for ethash caches dir=/root/.laser/testnet/glaze/ethash count=3
INFO [08-09|11:27:28.084] Disk storage enabled for ethash DAGs dir=/root/.laserhash count=2
INFO [08-09|11:27:28.084] Initialising Laser protocol versions="[63 62]" network=2
INFO [08-09|11:27:28.085] Loaded most recent local header number=27223 hash=1da4fc…84d3bc td=24994466404
INFO [08-09|11:27:28.085] Loaded most recent local full block number=27223 hash=1da4fc…84d3bc td=24994466404
INFO [08-09|11:27:28.085] Loaded most recent local fast block number=27223 hash=1da4fc…84d3bc td=24994466404
INFO [08-09|11:27:28.085] Loaded local transaction journal transactions=0 dropped=0
INFO [08-09|11:27:28.085] Regenerated local transaction journal transactions=0 accounts=0
WARN [08-09|11:27:28.086] Blockchain not empty, fast sync disabled
INFO [08-09|11:27:28.086] Starting P2P networking
INFO [08-09|11:27:30.320] UDP listener up self=enode://8eb9911c3149f2750152ac00cc1c7a5b3d7cb85a167eec86db3a49182b3f4ca99899ce988186f96508453eaa3bab5311a47c8293d65f42477dbc3853c7a00011@[::]:30503
INFO [08-09|11:27:30.320] RLPx listener up self=enode://8eb9911c3149f2750152ac00cc1c7a5b3d7cb85a167eec86db3a49182b3f4ca99899ce988186f96508453eaa3bab5311a47c8293d65f42477dbc3853c7a00011@[::]:30503
INFO [08-09|11:27:30.325] IPC endpoint opened url=/root/.laser/testnet/glaze.ipc
Connected to IPC socket
Exception in thread Thread-1:
Traceback (most recent call last):
File "threading.py", line 914, in _bootstrap_inner
File "src/nodeipc.py", line 100, in run
File "src/nodeipc.py", line 417, in handle_request
File "src/nodeipc.py", line 478, in handle_get_accounts
File "web3/personal.py", line 24, in listAccounts
File "web3/manager.py", line 107, in request_blocking
File "web3/manager.py", line 90, in _make_request
File "web3/middleware/gas_price_strategy.py", line 18, in middleware
File "cytoolz/functoolz.pyx", line 232, in cytoolz.functoolz.curry.call
File "web3/middleware/formatting.py", line 50, in apply_formatters
File "web3/middleware/attrdict.py", line 18, in middleware
File "cytoolz/functoolz.pyx", line 232, in cytoolz.functoolz.curry.call
File "web3/middleware/formatting.py", line 57, in apply_formatters
File "cytoolz/functoolz.pyx", line 232, in cytoolz.functoolz.curry.call
File "eth_utils/functional.py", line 22, in inner
File "web3/utils/formatters.py", line 79, in apply_formatter_to_array
File "eth_utils/address.py", line 114, in to_checksum_address
File "eth_utils/crypto.py", line 9, in keccak
File "eth_hash/main.py", line 38, in call
File "eth_hash/main.py", line 22, in _hasher_first_run
File "eth_hash/backends/auto.py", line 18, in keccak256
File "eth_hash/backends/auto.py", line 11, in _initialize
File "eth_hash/utils.py", line 15, in auto_choose_backend
File "eth_hash/utils.py", line 52, in choose_available_backend
ImportError: None of these hashing backends are installed: ['pycryptodome', 'pysha3'].
Install with pip install eth-hash[pycryptodome]

**packages mentioned in ImportError are already installed:**

root@hscuser-OptiPlex-3050:/home/hscuser/package# pip install pycryptodome
Requirement already satisfied: pycryptodome in /usr/local/lib/python3.5/dist-packages (3.6.4)
root@hscuser-OptiPlex-3050:/home/hscuser/package# pip install pysha3
Requirement already satisfied: pysha3 in /usr/local/lib/python3.5/dist-packages (1.0.2)

Any hint? whats going wrong?

**Comment by Shaptic** [comment]
I ran into this problem as well. Solved it by passing the following to PyInstaller: `--hidden-import=eth_hash.backends.pycryptodome`.

This is necessary because `eth_hash` does some weird patching of the import paths (using `importlib`) depending on the backends available, which PyInstaller doesn't pick up on.

**Comment by sheluchin** [comment]
@Shaptic Thanks! It looks like it's because:

>Python scripts import modules in ways that PyInstaller cannot detect: for example, by using the `__import__()` function with variable data, using `imp.find_module()`, or manipulating the `sys.path` value at run time. 

and eth-hash does exactly that here: https://github.com/ethereum/eth-hash/blob/d05c8ac0710e2dc16aae2c8a1e406ee7cbe78871/eth_hash/utils.py#L22-L23

Curiously enough, including `pycryptodome` didn't solve the issue for me, while `--hidden-import=eth_hash.backends.pysha3` did the trick.

**Comment by bjones1** [comment]
I've created a hook for this based on the information @Shaptic and @sheluchin provided. Would yall test it out? I don't use these modules, so I'm not able to test. This is `hook-eth_hash.py`:

``` Python
#-----------------------------------------------------------------------------
# Copyright (c) 2018-2018, PyInstaller Development Team.
#
# Distributed under the terms of the GNU General Public License with exception
# for distributing bootloader.
#
# The full license is in the file COPYING.txt, distributed with this software.
#-----------------------------------------------------------------------------

from PyInstaller.utils.hooks import collect_submodules

# The ``eth_hash.utils.load_backend`` function does a dynamic import.
hiddenimports = collect_submodules('eth_hash.backends')
```

**Comment by sheluchin** [comment]
@bjones1 Thanks, that does work. Is the best approach to just create a dir for hooks and reference it like `pyinstaller --additional-hooks-dir=pyinstaller_hooks foo.py`?

**Comment by bjones1** [comment]
Thanks for testing. I've just pushed those changes to the pyinstaller repo. To use, install the dev version.

Fixed in 0606c9d3beb5f90a6c4a64844d9e47ba402d5c97.
